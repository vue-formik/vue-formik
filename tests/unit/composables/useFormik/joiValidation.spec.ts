import { describe, test, expect } from "vitest";
import Joi from "joi";
import { useFormik } from "@/index";
import { nextTick } from "vue";

describe("joiValidation", async () => {
  const initialValues = {
    name: "",
    email: "",
    password: "",
    confirmPassword: "",
  };
  const validationSchema = Joi.object({
    name: Joi.string().required(),
    email: Joi.string().email().required(),
    password: Joi.string().min(6).required(),
    confirmPassword: Joi.string().valid(Joi.ref("password")).required(),
  });

  test("validate simple form", async () => {
    const formik = useFormik({
      initialValues,
      validationSchema,
      validateOnMount: true,
      mode: "JOI",
    });

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("name", "John Doe");
    formik.setFieldValue("email", "john@example.com");
    formik.setFieldValue("password", "password");
    formik.setFieldValue("confirmPassword", "password");

    await nextTick();
    expect(formik.errors).toMatchSnapshot();
    expect(formik.isValid.value).toBe(true);

    formik.setFieldValue("confirmPassword", "password123");
    await nextTick();
    expect(formik.errors).toMatchSnapshot();
    expect(formik.isValid.value).toBe(false);
  });

  test("validate form with array fields", async () => {
    const initialValues = {
      name: "",
      skills: [""],
    }
    const validationSchema = Joi.object({
      name: Joi.string().required(),
      skills: Joi.array().items(Joi.string().required()).required(),
    });

    const formik = useFormik({
      initialValues,
      validationSchema,
      validateOnMount: true,
      mode: "JOI",
    });

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("name", "John Doe");
    formik.setFieldValue("skills", ["Vue", "React"]);

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("skills", ["Vue", ""]);
    await nextTick();
    expect(formik.errors).toMatchSnapshot();
  });

  test("validate form with object fields", async () => {
    const initialValues = {
      name: "",
      address: {
        city: "",
        country: "",
      },
    }
    const validationSchema = Joi.object({
      name: Joi.string().required(),
      address: Joi.object({
        city: Joi.string().required(),
        country: Joi.string().required(),
      }).required(),
    });

    const formik = useFormik({
      initialValues,
      validationSchema,
      validateOnMount: true,
      mode: "JOI",
    });

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("name", "John Doe");
    formik.setFieldValue("address.city", "New York");
    formik.setFieldValue("address.country", "USA");

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("address.city", "");
    await nextTick();
    expect(formik.errors).toMatchSnapshot();
  });

  test("validate form with array of objects", async () => {
    const initialValues = {
      name: "",
      friends: [
        { name: "", email: "" },
      ],
    }
    const validationSchema = Joi.object({
      name: Joi.string().required(),
      friends: Joi.array().items(
        Joi.object({
          name: Joi.string().required(),
          email: Joi.string().email().required(),
        }).required()
      ).required(),
    });

    const formik = useFormik({
      initialValues,
      validationSchema,
      validateOnMount: true,
      mode: "JOI",
    });

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("name", "John Doe");
    formik.setFieldValue("friends[0].name", "Jane Doe");
    formik.setFieldValue("friends[0].email", "jane@ex.com");
    formik.setFieldValue("friends[1].name", "John Doe");
    formik.setFieldValue("friends[1].email", "john@ex.com");

    await nextTick();
    expect(formik.errors).toMatchSnapshot();

    formik.setFieldValue("friends[1].email", "john@ex");
    await nextTick();

    expect(formik.errors).toMatchSnapshot();
  })
});
