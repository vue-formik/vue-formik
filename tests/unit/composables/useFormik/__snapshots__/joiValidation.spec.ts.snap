// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`joiValidation > validate form with array fields 1`] = `
{
  "name": ""name" is not allowed to be empty",
  "skills": ""skills" does not contain 1 required value(s)",
}
`;

exports[`joiValidation > validate form with array fields 2`] = `{}`;

exports[`joiValidation > validate form with array fields 3`] = `
{
  "skills": [
    {},
    ""skills[1]" is not allowed to be empty",
  ],
}
`;

exports[`joiValidation > validate form with array of objects 1`] = `
{
  "friends": ""friends" does not contain 1 required value(s)",
  "name": ""name" is not allowed to be empty",
}
`;

exports[`joiValidation > validate form with array of objects 2`] = `{}`;

exports[`joiValidation > validate form with array of objects 3`] = `
{
  "friends": [
    {},
    {
      "email": ""friends[1].email" must be a valid email",
    },
  ],
}
`;

exports[`joiValidation > validate form with object fields 1`] = `
{
  "address": {
    "city": ""address.city" is not allowed to be empty",
    "country": ""address.country" is not allowed to be empty",
  },
  "name": ""name" is not allowed to be empty",
}
`;

exports[`joiValidation > validate form with object fields 2`] = `{}`;

exports[`joiValidation > validate form with object fields 3`] = `
{
  "address": {
    "city": ""address.city" is not allowed to be empty",
  },
}
`;

exports[`joiValidation > validate simple form 1`] = `
{
  "email": ""email" is not allowed to be empty",
  "name": ""name" is not allowed to be empty",
  "password": ""password" is not allowed to be empty",
}
`;

exports[`joiValidation > validate simple form 2`] = `{}`;

exports[`joiValidation > validate simple form 3`] = `
{
  "confirmPassword": ""confirmPassword" must be [ref:password]",
}
`;
